src/bin/train.cc

```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
         trainer->update(lr_decay);
 
         num_trained_samples += batch.source_ids[0].size();
+        num_trained_words += batch.target_ids.size() * batch.target_ids[0].size();
         if (!train_sampler.hasSamples()) {
           train_sampler.rewind();
         }
 
-        const string fmt_str = "Trained: batch=%d samples=%d lr=%.6e";
+        const string fmt_str = "Trained: batch=%d samples=%d current-proc-words=%d lr=%.6e";
         const auto fmt = boost::format(fmt_str)
-            % iteration % num_trained_samples % lr_decay;
+            % iteration % num_trained_samples % num_trained_words % lr_decay;
         logger->info(fmt.str());
       }
 
       if (lr_decay_type == "batch") {
         lr_decay *= lr_decay_ratio;
       }
 
-      if (iteration % eval_interval == 0) {
+      if (num_trained_samples >= next_epoch_samples) {
```

Comment:
This condition looks equal to !train_sampler.hasSamples() and below lines could be replaced to above if-statement (line 484).